File: Backtracking/combinationsum.java
class Solution {
    List<List<Integer>> res = new ArrayList();

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        backTrack(candidates, target, 0, new ArrayList());
        return res;
    }

    public void backTrack(int[] candidates, int target, int rejected, ArrayList<Integer> list) {
        if (target < 0)
            return;
        if (target == 0) {
            res.add(new ArrayList(list));
            return;
        }

        for (int i = rejected; i < candidates.length; i++) {
            list.add(candidates[i]);
            // chek if its taking me to a desired solution
            backTrack(candidates, target - candidates[i], i, list);
            list.remove(list.size() - 1); // actual backtracking
        }
    }
}

File: Backtracking/combinationsum.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []

        def dfs(i, curr, total):
            if (total == target):
                res.append(curr.copy())
                return

            if (i >= len(nums) or total > target):
                return

            dfs(i + 1, curr, total)
            curr.append(nums[i])
            dfs(i, curr, total + nums[i])
            curr.pop()
        
        dfs(0, [], 0)
        return res